File: computer_science/data_structures/graph/graph.py
class Graph:
    def __init__(self):
        self.vertices = []

    def add_vertex(self, key):
        if self.contains(self.vertices, key):
            print("Vertex", key, "not added because it is an existing key")
        else:
            self.vertices.append(Vertex(key))

    def print(self):
        for vertex in self.vertices:
            print("Vertex", vertex.key)
            for neighbor in vertex.adjacent:
                print(neighbor.key)

    def add_edge(self, from_key, to_key):
        from_vertex = self.get_vertex(from_key)
        to_vertex = self.get_vertex(to_key)

        if from_vertex is None or to_vertex is None:
            print("Invalid edge (", from_key, "-->", to_key, ")")
            return
        
        if self.contains(from_vertex.adjacent, to_key):
            print("Edge", to_key, "not added because it is already an existing neighbor")
            return

        from_vertex.adjacent.append(to_vertex)

    def get_vertex(self, key):
        for vertex in self.vertices:
            if vertex.key == key:
                return vertex

    def contains(self, vertices, key):
        for vertex in vertices:
            if vertex.key == key:
                return True

        return False

class Vertex:
    def __init__(self, key):
        self.key = key
        self.adjacent = []

File: computer_science/data_structures/graph/graph.js
class Vertex {
  constructor(key) {
    this.key = key;
    this.adjacent = [];
  }

  contains(key) {
    for (let vertex of this.adjacent) {
      if (vertex.key === key) {
        return true;
      }
    }

    return false;
  }
}

class Graph {
  constructor() {
    this.vertices = [];
  }

  addVertex(key) {
    if (this.contains(key)) {
      console.log(`Vertex ${key} not added because it's an existent key`);
      return;
    }

    this.vertices.push(new Vertex(key));
  }

  contains(key) {
    for (let vertex of this.vertices) {
      if (vertex.key === key) {
        return true;
      }
    }

    return false;
  }

  addEdge(from, to) {
    let fromVertex = this.getVertex(from);
    let toVertex = this.getVertex(to);

    if (!fromVertex) {
      console.log(`Vertex ${from} doesn't exist in the graph`);
      return;
    }

    if (!toVertex) {
      console.log(`Vertex ${to} doesn't exist in the graph`);
      return;
    }

    if (fromVertex.contains(to)) {
      console.log(`Existing edge: ${from} --> ${to}`);
      return;
    }

    fromVertex.adjacent.push(toVertex);
  }

  getVertex(key) {
    for (let vertex of this.vertices) {
      if (vertex.key === key) {
        return vertex;
      }
    }
  }

  print() {
    for (let vertex of this.vertices) {
      let adjacent = vertex.adjacent.reduce(
        (verticesString, vertex) => `${verticesString} ${vertex.key}`,
        ''
      );
      console.log(`Vertice ${vertex.key}: ${adjacent}`);
    }
  }
}

const graph = new Graph();

graph.addVertex(0);

for (let i = 1; i <= 4; i++) {
  graph.addVertex(i);
}

graph.addVertex(0);
graph.addVertex(0);

graph.addEdge(1, 2);
graph.addEdge(1, 3);
graph.addEdge(2, 3);
graph.addEdge(2, 4);
graph.addEdge(1, 2);
graph.addEdge(6, 2);
graph.addEdge(3, 2);
graph.addEdge(1, 1);

console.log();
graph.print();
